### ISSUE TITLE
Add PhotoInfo.owner for shared photos

### ISSUE THREAD

**Issue by RhetTbull** [RhetTbull/osxphotos#239]
`ZGENERICASSET.ZCLOUDOWNERHASHEDPERSONID` points to `ZCLOUDSHAREDALBUMINVITATIONRECORD.ZINVITEEHASHEDPERSONID`

**Comment by RhetTbull** [comment]
`ZGENERICALBUM.ZCLOUDOWNERFULLNAME`, `ZCLOUDOWNERHASHEDPERSONID`

**Comment by RhetTbull** [comment]
`ZGENERICASSET.ZSHAREORIGINATOR` might be helpful as well

**Comment by RhetTbull** [comment]
This will get the person's full name for shared photo with a given UUID

```sql
SELECT ZGENERICASSET.ZUUID, ZCLOUDSHAREDALBUMINVITATIONRECORD.ZINVITEEFULLNAME, ZCLOUDSHAREDALBUMINVITATIONRECORD.ZINVITEEHASHEDPERSONID
FROM ZCLOUDSHAREDALBUMINVITATIONRECORD 
JOIN ZGENERICASSET ON ZGENERICASSET.ZCLOUDOWNERHASHEDPERSONID = ZCLOUDSHAREDALBUMINVITATIONRECORD.ZINVITEEHASHEDPERSONID
WHERE ZGENERICASSET.ZUUID = "83EE2759-6875-43A6-99C8-184FBE51AC9C"
ORDER BY ZCLOUDSHAREDALBUMINVITATIONRECORD.Z_PK
LIMIT 1
```

**Comment by RhetTbull** [comment]
I have a beta implementation of this working.  Just need to add tests.  I've added a new way to easily add new properties to PhotoInfo without having to modify the code in photosdb which is very ugly and complex.  Basically, added a way to execute a sql query against the one property you're interested in which makes it much easier to add new properties like this in the future.  For example, the ZCLOUDSHAREDALBUMINVITATIONRECORD has a "ZISMINE" column which is true if the photo is yours versus someone else's that was shared with you.  This would be useful in addition to the owner name.

**Comment by RhetTbull** [comment]
So this works for photos that people have shared to albums you invited them to but no for photos where someone else invited you to the shared album.

In the latter case, it looks like I'll need to get name from ZGENERICALBUM where ZGENERICALBUM is JOINed on ZCLOUDOWNERHASHEDPERSONID

**Comment by RhetTbull** [comment]
For the latter case, this looks like it'll work:

```sql
SELECT ZGENERICALBUM.ZCLOUDOWNERFULLNAME
FROM ZGENERICALBUM
JOIN ZGENERICASSET ON ZGENERICASSET.ZCLOUDOWNERHASHEDPERSONID = ZGENERICALBUM.ZCLOUDOWNERHASHEDPERSONID
WHERE ZGENERICASSET.ZUUID = "9AAA73E1-1BCC-431D-BFD2-1D2C4A037A81"
```

**Comment by RhetTbull** [comment]
It would be good to have a way to do this in a single SQL query but that may be beyond my SQL abilities

**Comment by RhetTbull** [comment]
I think this will work:

```sql
WITH case1 AS
(
SELECT ZGENERICALBUM.ZCLOUDOWNERFULLNAME as OWNER_FULLNAME
FROM ZGENERICALBUM
JOIN ZGENERICASSET ON ZGENERICASSET.ZCLOUDOWNERHASHEDPERSONID = ZGENERICALBUM.ZCLOUDOWNERHASHEDPERSONID
WHERE ZGENERICASSET.ZUUID = "983B75C9-6991-41A7-9DCE-43E9DAA15C28"
),
case2 AS
(
SELECT
ZCLOUDSHAREDALBUMINVITATIONRECORD.ZINVITEEFULLNAME AS OWNER_FULLNAME
FROM ZCLOUDSHAREDALBUMINVITATIONRECORD 
JOIN ZGENERICASSET ON ZGENERICASSET.ZCLOUDOWNERHASHEDPERSONID = ZCLOUDSHAREDALBUMINVITATIONRECORD.ZINVITEEHASHEDPERSONID
WHERE ZGENERICASSET.ZUUID = "983B75C9-6991-41A7-9DCE-43E9DAA15C28"
ORDER BY ZCLOUDSHAREDALBUMINVITATIONRECORD.Z_PK
LIMIT 1
)
SELECT * FROM case1
UNION
SELECT * FROM case2 WHERE NOT EXISTS (SELECT * FROM case1)
```

**Comment by RhetTbull** [comment]
`PhotoInfo.owner` (and `AlbumInfo.owner`) were implemented in v0.42.85.  I tried a new way of implementing photo properties to make the code more maintainable but after doing some benchmarking, not sure about this as a long term solution.  

osxphotos works but querying the sqlite database that Photos uses to store data about the library.  Normally, all the data needed by osxphotos is read at startup (when the `PhotosDB()` class is instantiated) and stored in in-memory data structures (python `dict`) which are then accessed by PhotoInfo and other classes as needed to retrieve the information.  This means the `PhotosDB` class is quite complex and I've got to keep track of all the various data structures.  

For implementing the `owner` property,  I added the ability to [execute a query](https://github.com/RhetTbull/osxphotos/blob/c86e84c534424859d397387c764307aca6f39f42/osxphotos/photosdb/photosdb.py#L3361-L3363)  against the database and return the results as well as the ability to load a [sql query from disk](https://github.com/RhetTbull/osxphotos/blob/master/osxphotos/query_builder.py) as a [template](https://github.com/RhetTbull/osxphotos/blob/master/osxphotos/queries/shared_owner.sql.mako).  This makes the implementation of `owner` and future properties very easy as one doesn't need to understand the complex and ugly code in `PhotosDB` to add new features--you just need to craft a SQL query against the Photos database, the structure of which is documented in the [osxphotos wiki](https://github.com/RhetTbull/osxphotos/wiki).  

Unfortunately, I didn't realize how slow this would be.  For example, against my library of 30536 photos, this means looking up `owner` makes 30536 sqlite queries and this is *slow*.   I've taken steps to speed it up -- e.g. the SQL query that's loaded of disk is memoized so it only gets loaded once, the result is memoized so repeated lookups of `owner` don't have to hit the database, etc but it's still really slow.  Benchmarking results follow. Not sure if I should keep this implementation as this may not be a frequently used property or refactor this to load all the `owner` info at startup along with the other pre-loaded data.  The advantage here is it can be done in a single query instead of 30k+ separate queries.

```
python bench.py
Loading database
Getting photos
Got 30536 photos
Getting photos with title
Got 62 in 0.025536060333251953 seconds
Getting photos with owner
Got 6350 in 108.28019785881042 seconds
Getting photos with owner again
Got 6350 in 0.06004810333251953 seconds
```

bench.py:
```python
import osxphotos
import time

print(f"Loading database")
db = osxphotos.PhotosDB()

print(f"Getting photos")
photos = db.photos(images=True, movies=True)
print(f"Got {len(photos)} photos")

print(f"Getting photos with title")
start_t = time.time()
new_photos = [p for p in photos if p.title]
stop_t = time.time()
print(f"Got {len(new_photos)} in {stop_t-start_t} seconds")

print(f"Getting photos with owner")
start_t = time.time()
new_photos = [p for p in photos if p.owner]
stop_t = time.time()
print(f"Got {len(new_photos)} in {stop_t-start_t} seconds")

print(f"Getting photos with owner again")
start_t = time.time()
new_photos = [p for p in photos if p.owner]
stop_t = time.time()
print(f"Got {len(new_photos)} in {stop_t-start_t} seconds")
```

**Comment by RhetTbull** [comment]
Also, the current implementation is not 100% correct. It returns "(null) (null)" for some photos.

**Comment by RhetTbull** [comment]
Looks like pre-loading the owner info is definitely the way to go.  Simple test below shows that all owner info can be pre-loaded in 0.25 seconds (on my 2013 macbook pro, it already takes ~15 seconds for osxphotos to load the entire database so this is negligible):

```
Loading database
Getting photos
Got 30536 photos
Loading owner info
owners = 2365
owners = 6350
Got 6350 owner records in 0.2652130126953125 seconds
```

```python
import osxphotos
import time

print(f"Loading database")
db = osxphotos.PhotosDB()

print(f"Getting photos")
photos = db.photos(images=True, movies=True)
print(f"Got {len(photos)} photos")

owners = {}
print(f"Loading owner info")
sql_owner_case_1 = """
    SELECT DISTINCT ZGENERICASSET.ZUUID as UUID, ZGENERICALBUM.ZCLOUDOWNERFULLNAME as OWNER_FULLNAME
    FROM ZGENERICALBUM
    JOIN ZGENERICASSET ON ZGENERICASSET.ZCLOUDOWNERHASHEDPERSONID = ZGENERICALBUM.ZCLOUDOWNERHASHEDPERSONID
    WHERE OWNER_FULLNAME != '(null) (null)'
"""
sql_owner_case_2 = """
    SELECT DISTINCT
    ZGENERICASSET.ZUUID AS UUID, ZCLOUDSHAREDALBUMINVITATIONRECORD.ZINVITEEFULLNAME AS OWNER_FULLNAME
    FROM ZCLOUDSHAREDALBUMINVITATIONRECORD 
    JOIN ZGENERICASSET ON ZGENERICASSET.ZCLOUDOWNERHASHEDPERSONID = ZCLOUDSHAREDALBUMINVITATIONRECORD.ZINVITEEHASHEDPERSONID
"""
start_t = time.time()

results = db.execute(sql_owner_case_1)
for row in results:
    owners[row[0]] = row[1]

print(f"owners = {len(owners)}")

results = db.execute(sql_owner_case_2)
for row in results:
    owners[row[0]] = row[1]
print(f"owners = {len(owners)}")
stop_t = time.time()
print(f"Got {len(owners)} owner records in {stop_t-start_t} seconds")
```

**Comment by RhetTbull** [comment]
This is more like it...new implementation that pre-loads the necessary data

```
>>> import time
>>> start_t = time.time(); owners = [(p.uuid, p.owner) for p in photos if p.owner]; stop_t = time.time()
>>> stop_t - start_t
0.07246017456054688
>>> len(owners)
6350
>>>
```

**Comment by RhetTbull** [comment]
Fixed in [v0.42.88](https://github.com/RhetTbull/osxphotos/releases/tag/v0.42.88).  Still need to apply the same fix to album.owner

**Comment by RhetTbull** [comment]
Fixed (again) in [v0.42.89](https://github.com/RhetTbull/osxphotos/releases/tag/v0.42.89)
